### PR TITLE
BUG: Keep column level names during resample nunique (#23222)

### LINKED ISSUE #23222: Resampling using `nunique` causes multi-level columns to lose their level names
#### Problem description

Resampling using `nunique` causes multi-level columns to lose their level names.
https://nbviewer.jupyter.org/gist/taljaards/20e945b7572aea1f4eb4aa4c6e823037

I only ran into this issue with `nunique`; I do not know if this is also the case for some other functions.

#### Expected Output
To not drop the level names, as in the first resample example.

#### Output of ``pd.show_versions()``

<details>

```
import pandas as pd
pd.show_versions()
Backend TkAgg is interactive backend. Turning interactive mode on.
Matplotlib support failed
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\JetBrains\Toolbox\apps\PyCharm-P\ch-0\183.3647.8\helpers\pydev\_pydev_bundle\pydev_import_hook.py", line 23, in do_import
    succeeded = activate_func()
  File "C:\Users\Admin\AppData\Local\JetBrains\Toolbox\apps\PyCharm-P\ch-0\183.3647.8\helpers\pydev\pydev_ipython\matplotlibtools.py", line 141, in activate_pylab
    pylab = sys.modules['pylab']
KeyError: 'pylab'
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.6.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 142 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None
pandas: 0.23.4
pytest: 3.8.2
pip: 18.1
setuptools: 40.4.3
Cython: 0.28.5
numpy: 1.15.1
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 7.0.1
sphinx: 1.8.1
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.8
feather: None
matplotlib: 3.0.0
openpyxl: 2.5.8
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.1.1
lxml: 4.2.5
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: 1.2.12
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
```
</details>

### PR DESCRIPTION
**Pull request by dsm054**
- [X] closes #23222
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

nunique is the only one I found in a quick search which manifested the problem.

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.txt [hunk 1]
```diff
@@ -1335,6 +1335,7 @@ Groupby/Resample/Rolling
 - Bug in :meth:`DataFrame.resample` and :meth:`Series.resample` when resampling by a weekly offset (``'W'``) across a DST transition (:issue:`9119`, :issue:`21459`)
 - Bug in :meth:`DataFrame.expanding` in which the ``axis`` argument was not being respected during aggregations (:issue:`23372`)
 - Bug in :meth:`pandas.core.groupby.DataFrameGroupBy.transform` which caused missing values when the input function can accept a :class:`DataFrame` but renames it (:issue:`23455`).
+- Bug in :meth:`pandas.core.groupby.DataFrameGroupBy.nunique` in which the names of column levels were lost (:issue:`23222`).
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.txt]: Looks like `DataFrameGroupBy.nunique` isn't in `api.rst`. Could you add it?

File: pandas/tests/groupby/test_function.py [hunk 2]
```diff
@@ -892,6 +892,14 @@ def test_nunique_with_timegrouper():
     tm.assert_series_equal(result, expected)
 
 
+def test_nunique_preserves_column_level_names():
+    # GH 23222
+    test = pd.DataFrame([1, 2, 3],
+                        columns=pd.Index(['A'], name="level_0"))
+    result = test.groupby([0, 0, 0]).nunique()
+    tm.assert_index_equal(test.columns, result.columns)
```
Comment (gfyoung) [inline pandas/tests/groupby/test_function.py]: I know the bug is focused on the names but for robustness, let's compare the entire `DataFrame`.

Comment (dsm054) [reply pandas/tests/groupby/test_function.py]: Heh.  I attended a talk on the weekend where the speaker argued for single-purpose tests and against byproduct testing. ;-)  But it's easy to change, and I'm happy to respect house style.

Comment (gfyoung) [reply pandas/tests/groupby/test_function.py]: Well, consider your single purpose to check that the output of `nunique` is correct 😉.  Thanks!

**Review by gfyoung — APPROVED** [review]

**Comment by jreback** [comment]
if you can merge master

**Comment by jreback** [comment]
@dsm054 can you merge master

**Comment by WillAyd** [comment]
Closing as stale. Ping if you'd like to continue

**Comment by gfyoung** [comment]
Revived in #25469

**Pull request closed without merging**